### PR TITLE
Implement bswap in SAT back-end

### PR DESCRIPTION
**Pull request by tautschnig**
Previously only constants were supported via expression simplification.

### CODE REVIEW

File: regression/cbmc/gcc_bswap1/main.c [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  assert(__builtin_bswap16(sa)==0xcdab);
+  unsigned short sb;
+  sb&=0xff00;
+  assert((__builtin_bswap16(sb)&0xff00)==0);
+
+  unsigned a=0xabcdef00;
+  assert(__builtin_bswap32(a)==0x00efcdab);
+  unsigned b;
+  b&=0xffffff00;
+  assert((__builtin_bswap32(b)&0xff000000)==0);
```
Comment (smowton) [inline regression/cbmc/gcc_bswap1/main.c]: Note GCC has bswap64 as well, might as well get that while we're here?

File: src/solvers/flattening/boolbv_bswap.cpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+bvt boolbvt::convert_bswap(const exprt &expr)
+{
+  if(expr.id()!=ID_bswap || expr.operands().size()!=1)
+    throw "expected bswap with one operand";
+
+  std::size_t width=boolbv_width(expr.type());
+
+  // width must be multiple of bytes
+  if(width%config.ansi_c.char_width!=0)
```
Comment (smowton) [inline src/solvers/flattening/boolbv_bswap.cpp]: Now that operator spacing is allowed in the CBMC codebase, let's use it to improve readability :)

**Comment by tautschnig** [comment]
@smowton Thanks a lot for the comments, all addressed.

File: src/solvers/flattening/boolbv_bswap.cpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+\*******************************************************************/
+
+#include "boolbv.h"
+
+#include <util/config.h>
+#include <util/invariant.h>
+
+bvt boolbvt::convert_bswap(const exprt &expr)
+{
+  if(expr.id() != ID_bswap || expr.operands().size() != 1)
```
Comment (romainbrenguier) [inline src/solvers/flattening/boolbv_bswap.cpp]: Having a precondition seems more appropriate

File: src/solvers/flattening/boolbv_bswap.cpp [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    throw "expected bswap with one operand";
+
+  std::size_t width = boolbv_width(expr.type());
```
Comment (romainbrenguier) [inline src/solvers/flattening/boolbv_bswap.cpp]: should be const

File: src/solvers/flattening/boolbv_bswap.cpp [hunk 5]
```diff
@@ -0,0 +1,43 @@
+/*******************************************************************\
+
+Module:
```
Comment (romainbrenguier) [inline src/solvers/flattening/boolbv_bswap.cpp]: I think we should have a module name or description here

**Review by smowton — APPROVED** [review]

File: src/solvers/flattening/boolbv_bswap.cpp (hunk 3 above)
Comment (romainbrenguier) [inline src/solvers/flattening/boolbv_bswap.cpp]: I don't like that we can have expressions with their own id and no associated class (inheriting from `exprt`). In particular it is not documented anywhere how these bswap expressions are structured.

File: regression/cbmc/gcc_bswap1/test.desc [hunk 6]
```diff
@@ -0,0 +1,8 @@
+CORE
+main.c
+
+^EXIT=0$
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
```
Comment (romainbrenguier) [inline regression/cbmc/gcc_bswap1/test.desc]: Having one positive test does not seem enough to show that it works (a contradiction in the solver would give the same result).
We should have also some tests for which verification fail (for instance bswap a particular value and assert it is different from the expected result).

**Review by romainbrenguier — CHANGES_REQUESTED** [review]

**Comment by tautschnig** [comment]
@romainbrenguier Thank you for pushing me to do the cleanup! I believe I have addressed all the issues you raised.

File: regression/cbmc/gcc_bswap1/main.c [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+uint16_t __builtin_bswap16(uint16_t);
+uint32_t __builtin_bswap32(uint32_t);
+uint64_t __builtin_bswap64(uint64_t);
+#endif
+
+int main()
+{
+  uint16_t sa = 0xabcd;
+  assert(__builtin_bswap16(sa) == 0xcdab);
+  uint16_t sb;
```
Comment (martin-cs) [inline regression/cbmc/gcc_bswap1/main.c]: Pedantry but, no chance of a nondet_whatever() function call for the things that should be nondet.  Doesn't change behaviour but helps people see what is going on.

File: src/solvers/flattening/boolbv_bswap.cpp [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+{
+  const std::size_t width = boolbv_width(expr.type());
+
+  // width must be multiple of bytes
+  const std::size_t byte_bits = config.ansi_c.char_width;
+  if(width % byte_bits != 0)
+    return conversion_failed(expr);
+
+  bvt result = convert_bv(expr.op());
+  CHECK_RETURN(result.size() == width);
```
Comment (martin-cs) [inline src/solvers/flattening/boolbv_bswap.cpp:24]: Tidy!  I also note that convert_bv is correct, convert_bitvector is not.

File: src/util/std_expr.h [hunk 9]
```diff
@@ -442,6 +442,78 @@ inline void validate_expr(const unary_minus_exprt &value)
...
+inline const bswap_exprt &to_bswap_expr(const exprt &expr)
+{
+  PRECONDITION(expr.id()==ID_bswap);
+  DATA_INVARIANT(
+    expr.operands().size()==1,
+    "bswap must have one operand");
+  DATA_INVARIANT(
+    expr.op0().type()==expr.type(),
+    "bswap type must match operand type");
+  return static_cast<const bswap_exprt &>(expr);
```
Comment (martin-cs) [inline src/util/std_expr.h]: DATA_INVARIANTs about types are a most welcome addition.  @chrisr-diffblue will approve!

**Review by martin-cs — APPROVED** [review]
LGTM.  Am willing to merge once @romainbrenguier says he is happy with the changes he requested.

**Comment by tautschnig** [comment]
@romainbrenguier @martin-cs I believe I've addressed all change suggestions (unless I misunderstood any of the above comments).

File: regression/cbmc/gcc_bswap1/test.desc [hunk 10]
```diff
@@ -0,0 +1,11 @@
+CORE
+main.c
+
+^EXIT=10$
+^SIGNAL=0$
+^VERIFICATION FAILED$
+\[main\.assertion\..\] assertion __builtin_bswap(16|32|64)\((sb|b|bl)\) == 0: FAILURE$
```
Comment (romainbrenguier) [inline regression/cbmc/gcc_bswap1/test.desc]: could you put the assertion number so that it is clear which ones are supposed to fail

Comment (tautschnig) [reply regression/cbmc/gcc_bswap1/test.desc]: Done.

**Review by romainbrenguier — APPROVED** [review]

**Comment by tautschnig** [comment]
Pinging @kroening for a review.

**Comment by tautschnig** [comment]
@kroening ping?

**Review by peterschrammel — APPROVED** [review]
Lgtm

**Comment by tautschnig** [comment]
Will still need @kroening's approval...

File: regression/cbmc/gcc_bswap1/main.c [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+uint64_t __VERIFIER_nondet_u64();
+
+int main()
+{
+  uint16_t sa = 0xabcd;
+  assert(__builtin_bswap16(sa) == 0xcdab);
+  uint16_t sb = __VERIFIER_nondet_u16();
+  sb &= 0xff00;
+  // expected to fail, only MSB guaranteed to be 0
+  assert(__builtin_bswap16(sb) == 0);
```
Comment (hannes-steffenhagen-diffblue) [inline regression/cbmc/gcc_bswap1/main.c:21]: Shouldn't this be unknown instead of failing if it's unknown?

Comment (tautschnig) [reply regression/cbmc/gcc_bswap1/main.c:21]: Is that asking to clarify the comment (which should likely say "may fail")?

**Review by kroening — APPROVED** [review]

**Pull request merged**